Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs

```diff
@@ -1,74 +1,192 @@
+            if (user == null)
+            {
+                throw new SharedException(
+                    ErrorTypes.NotFound,
+                    "User not found",
+                    "User with such username not found"
+                );
+            }
+
+            return user;
         }
 
-        public bool ValidatePassword(User existingUser, UserLoginDTO userDTO)
+        public bool CheckIfPasswordValid(User existingUser, UserLoginDTO userLoginDTO)
         {
             return _passwordService.VerifyPassword(
                 existingUser.HashedPassword,
                 existingUser.Salt,
-                userDTO.Password
+                userLoginDTO.Password
             );
         }
 
-        public ValueTask<User?> GetByRefreshTokenAsync(RefreshToken token)
+        /// <exception cref="SharedException"><paramref name="token"/> subject not found.</exception>
+        public async Task<User> GetUserByRefreshTokenAsync(RefreshToken token)
         {
-            return _userRepository.GetByIdAsync(token.UserId);
+            var user = await _userDAO.GetByIdAsync(token.UserId);
+            if (user == null)
+            {
+                throw new SharedException(
+                    ErrorTypes.AuthFailed,
+                    "Invalid token",
+                    "User not found by refresh token userId"
+                );
+            }
+
+            return user;
         }
 
-        public async Task<User> RegisterAsync(UserRegistrationDTO userDTO)
+        public async Task<User> RegisterAsync(UserRegistrationDTO userRegistrationDTO)
         {
-            var user = ConvertFromDTO(userDTO);
+            var user = CreateFromDTO(userRegistrationDTO);
 
-            return await _userRepository.CreateUserAsync(user);
+            await _userDAO.InsertUserAsync(user);
+
+            return user;
         }
 
-        public Task<bool> CheckIfExistsAsync(UserRegistrationDTO userDTO)
+        public Task UpgradeRoleAsync(User user)
         {
-            return _userRepository.IsUniqueCredentialsAsync(
-                userDTO.Username, userDTO.Email
-            );
+            switch (user.Roles)
+            {
+                case Roles.None:
+                {
+                    user.Roles = Roles.Admin;
+                    break;
+                }
+                case Roles.Client:
+                {
+                    user.Roles = Roles.Admin;
+                    break;
+                }
+                case Roles.Admin:
+                {
+                    user.Roles = Roles.SuperAdmin;
+                    break;
+                }
+                case Roles.SuperAdmin:
+                {
+                    throw new SharedException(
+                        ErrorTypes.Conflict,
+                        "Role upgrading failed",
+                        "User already has highest role"
+                    );
+                }
```

Comment:
This is a bit complex and not intuitive role system implementation (I haven't seen UI, so can't tell if it's user friendly as well). The system that is used more often allows assigning any number of roles to a user and then checking if he has has that specific role for that particular action.

Reply:
Thank you for comment. We also discussed this point with Kseniya. And now implementation of suggested roles scheme is in progress.